release.nix: generate bootstrap tools for musl

alyssais opened the pull request:
## Description of changes

This is a fresh attempt at https://github.com/NixOS/nixpkgs/pull/169793, done in a way that should make @Ericson2314 happier — i.e., all `bootstrapTools` jobs are switched to use config triples.

This will allow us to regenerate our woefully out-of-date aarch64-unknown-linux-gnu musl bootstrap tools, which can't compile lots of modern code.

We could technically also do i686-unknown-linux-musl bootstrap tools, but I don't know if there's demand for that, so it's best to wait and see if somebody asks for it before we commit Hydra to it.

I've tested the switch to triples by running

    nix-eval-jobs --force-recurse pkgs/top-level/release.nix

both before and after this change, and diffing the results to verify that no new errors are introduced, the derivations for all bootstrap tools are the same, and that the only changes aside from the bootstrap tool jobs being renamed are to derivations that include the Nixpkgs source tree.

Closes #169793.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

alyssais committed:
release.nix: generate bootstrap tools for musl

This will allow us to regenerate our woefully out-of-date
aarch64-unknown-linux-gnu musl bootstrap tools, which can't compile
lots of modern code.

We could technically also do i686-unknown-linux-musl bootstrap tools,
but I don't know if there's demand for that, so it's best to wait and
see if somebody asks for it before we commit Hydra to it.

RaitoBezarius reviewed (approved):
Changes are reasonable, checked with a character diff before and after using `nix-eval-jobs` as suggested.

Ericson2314 commented on nixos/release.nix:
```diff
@@ -77,36 +77,37 @@ let
     in
       tarball //
         { meta = {
-            description = "NixOS system tarball for ${system} - ${stdenv.hostPlatform.linux-kernel.name}";
```
Is this `stdenv` still in scope?

Ericson2314 reviewed (approved):
Very excited for this! Just have on question

Ericson2314 commented on pkgs/top-level/release.nix:
```diff
@@ -11,8 +11,8 @@
   # The platforms for which we build Nixpkgs.
-, supportedSystems ? [ "x86_64-linux" "x86_64-darwin" "aarch64-linux" "aarch64-darwin" ]
-, limitedSupportedSystems ? [ "i686-linux" ]
+, supportedSystems ? [ "x86_64-unknown-linux-gnu" "x86_64-apple-darwin" "aarch64-unknown-linux-gnu" "aarch64-apple-darwin" ]
+, limitedSupportedSystems ? [ "aarch64-unknown-linux-musl" "i686-unknown-linux-gnu" "x86_64-unknown-linux-musl" ]
```
```suggestion
, supportedConfigs ? [ "x86_64-unknown-linux-gnu" "x86_64-apple-darwin" "aarch64-unknown-linux-gnu" "aarch64-apple-darwin" ]
, limitedSupportedConfigs ? [ "aarch64-unknown-linux-musl" "i686-unknown-linux-gnu" "x86_64-unknown-linux-musl" ]
```
I wonder if this helps or hurts

alyssais replied:
Hmm, yeah, we're expecting a different format, so renaming the parameters (hence making it visible to callers that something has changed) makes sense.

alyssais replied:
It actually looks like there are callers of this file that I forgot to update, so I'll fix that too.

vcunat replied:
Not in many cases, I think.  At hydra.nixos.org instance at least :-)

alyssais replied:
I've had a go at the first one, but it's too much work.  Maybe one day.  It would also have some other drawbacks, like renaming every single Hydra job.  I'll see how feasible the second one is.

vcunat replied:
Ah right, renaming jobs would be unpleasant.

Ericson2314 replied:
I don't mind the extra work part :P, but yes I suppose someone could get made if all the hydra job names start changing.

alyssais replied:
(No longer relevant to the new attempt.)

Ericson2314 reviewed (commented):
Very excited about this! Just a few small questions to try to make sure no one ends up trying to revert it later :)

alyssais committed:
release.nix: namespace bootstrap tools with triples

This will allow buliding bootstrap tools for platforms with
non-default libcs, like *-unknown-linux-musl.

This gets rid of limitedSupportSystems/systemsWithAnySupport.  There
was no need to use systemsWithAnySupport for supportDarwin, because it
was always equivalent to supportedSystems for that purpose, and the
only other way it was used was for determining which platforms to
build the bootstrap tools for, so we might as well use a more explicit
parameter for that, and then we can change how it works without
affecting the rest of the Hydra jobs.

Not affecting the rest of the Hydra jobs is important, because if we
changed all jobs to use config triples, we'd end up renaming every
Hydra job.  That might still be worth thinking about at some point,
but it's unnecessary at this point (and would be a lot of work).

I've checked by running

    nix-eval-jobs --force-recurse pkgs/top-level/release.nix

that the actual bootstrap tools derivations are unaffected by this
change, and that the only other jobs that change are ones that depend
on the hash of all of Nixpkgs.  Of the other jobset entrypoints that
end up importing pkgs/top-level/release.nix, none used the
limitedSupportedSystems parameter, so they should all be unaffected as
well.

alyssais commented:
I've now pushed a third attempt at this, that avoids the complexity the second attempt still hadn't figured out.

The problem in both previous attempts has been that converting everything to use config triples is a huge job, with various side effects like renaming almost every single Hydra job.

This new attempt recognises that the list of platforms with bootstrap tools built on Hydra was already distinct from the list of platforms everything else was built for, so it just makes that a bit clearer, and then changes just that list to config triples.  This way, we can achieve the goal of musl bootstrap tools built by Hydra, without doing a total overhaul of Hydra job generation in Nixpkgs

Ericson2314 reviewed (approved):
OK great, that's do this!
